File: ios/Plugin/UdpSocketPlugin.swift
import Foundation
import Capacitor
import CocoaAsyncSocket

public enum SocketsError: Error {
    case Error (String)
}

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(UdpSocketPlugin)
public class UdpSocketPlugin: CAPPlugin {
    private var sockets: [Int: UdpSocket] = [Int: UdpSocket]()
    private var nextSocketId: Int = 0

    override public func load() {
        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: "capacitor-udp-forward"), object: nil, queue: nil, using: handleUdpForward )
    }

    @objc func create(_ call: CAPPluginCall) {
        let properties = call.getObject("properties")

        let socket = UdpSocket.init(id: nextSocketId, properties: properties)
        sockets[socket.socketId] = socket
        socket.onReceivedHandler = { data in
            self.notifyListeners("receive", data: data, retainUntilConsumed: false)
        }

        socket.onReceivedErrorHandler = { data in
            self.notifyListeners("receiveError", data: data, retainUntilConsumed: false)
        }

        nextSocketId+=1

        call.resolve([
            "socketId": socket.socketId,
            "ipv4": Utils.getIPv4Address() ?? "",
            "ipv6": Utils.getIPv6Address() ?? ""
        ])
    }

    @objc func update(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let properties: [String: Any] = call.getObject("properties") ?? [String: Any]()
        socket.setProperties(properties)
        call.resolve()
    }

    @objc func setPaused(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let paused = call.getBool("paused", false)
        socket.setPaused(paused)
        call.resolve()
    }

    @objc func bind(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        guard let port = call.getInt("port") else {
            call.reject("Illegal port")
            return
        }
        let address = call.getString("address")
        do {
            try socket.bind(port, address: address)
            call.resolve()
        } catch let SocketsError.Error(msg) {
            call.reject(msg)
        } catch {
            call.reject("bind Error")
        }

    }

    @objc func send(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        guard let port = call.getInt("port") else {
            call.reject("Illegal port")
            return
        }

        let address = call.getString("address", "")
        let dataString = call.getString("buffer", "")

        do {
            let data = Data(base64Encoded: dataString, options: .ignoreUnknownCharacters) ?? Data.init()
            try socket.send(data, address: address, port: port)
            call.resolve(["bytesSent": data.count])
        } catch let SocketsError.Error(msg) {
            call.reject(msg)
        } catch {
            call.reject("unkow error")
        }
    }

    @objc func close(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        socket.closeSocket()
        sockets[socketId] = nil
        call.resolve()
    }

    @objc func closeAllSockets(_ call: CAPPluginCall) {
        for (socketId, socket) in sockets {
            socket.closeSocket()
            sockets[socketId] = nil
        }
        call.resolve([
            "success": "close all"
        ])
    }

    @objc func getInfo(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        call.resolve(socket.getInfo())
    }

    @objc func getSockets(_ call: CAPPluginCall) {
        var socketsInfo = [Any]()
        for (_, socket) in sockets {
            socketsInfo.append(socket.getInfo())
        }
        call.resolve(["sockets": socketsInfo])
    }

    @objc func setBroadcast(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let enabled = call.getBool("enabled", false)
        do {
            try socket.setBroadcast(enabled)
            call.resolve()
        } catch let SocketsError.Error(msg) {
            call.reject(msg)
        } catch {
            call.reject("unkow error")
        }
    }

    @objc func joinGroup(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let address = call.getString("address", "")
        do {
            try socket.joinGroup(address)
            call.resolve()
        } catch let SocketsError.Error(msg) {
            call.reject(msg)
        } catch {
            call.reject("unkow error")
        }
    }

    @objc func leaveGroup(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let address = call.getString("address", "")
        do {
            try socket.leaveGroup(address)
            call.resolve()
        } catch let SocketsError.Error(msg) {
            call.reject(msg)
        } catch {
            call.reject("unkow error")
        }

    }

    @objc func getJoinedGroups(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }

        let groups = socket.multicastGroup
        var groupArray = [String]()
        for group in groups {
            groupArray.append(group)
        }
        call.resolve(["groups": groupArray])
    }

    @objc func setMulticastTimeToLive(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }

        guard let ttl = call.getInt("ttl") else {
            call.reject("Illegal ttl")
            return
        }

        socket.socket?.perform({ () -> Void in
            if socket.socket?.isIPv4() ?? false {
                var ttlCpy: CUnsignedChar = (ttl as NSNumber).uint8Value
                if setsockopt(socket.socket?.socket4FD() ?? 0, IPPROTO_IP, IP_MULTICAST_TTL, &ttlCpy, UInt32( MemoryLayout.size(ofValue: ttlCpy))) < 0 {
                    call.reject("ttl ipv4 error")
                }
            }
            if socket.socket?.isIPv6() ?? false {
                var ttlCpy = ttl
                if setsockopt(socket.socket?.socket6FD() ?? 0, IPPROTO_IPV6, IP_MULTICAST_TTL, &ttlCpy, UInt32( MemoryLayout.size(ofValue: ttlCpy))) < 0 {
                    call.reject("ttl ipv6 error")
                }
            }
            call.resolve()
        })
    }

    @objc func setMulticastLoopbackMode(_ call: CAPPluginCall) {
        guard let socketId = call.getInt("socketId"), let socket = sockets[socketId] else {
            call.reject("Socket not found")
            return
        }
        let enabled = call.getBool("enabled", false)

        socket.socket?.perform({ () -> Void in
            if socket.socket?.isIPv4() ?? false {
                var loop: CUnsignedChar = enabled ? 1 : 0
                if setsockopt(socket.socket?.socket4FD() ?? 0, IPPROTO_IP, IP_MULTICAST_LOOP, &loop, UInt32( MemoryLayout.size(ofValue: loop))) < 0 {
                    call.reject("loopback ipv4 error")
                }
            }
            if socket.socket?.isIPv6() ?? false {
                var loop: Int32 = enabled ? 1 : 0
                if setsockopt(socket.socket?.socket6FD() ?? 0, IPPROTO_IPV6, IP_MULTICAST_LOOP, &loop, UInt32( MemoryLayout.size(ofValue: loop))) < 0 {
                    call.reject("loopback ipv6 error")
                }
            }
            call.resolve()
        })
    }

    private func handleUdpForward(_ notification: Notification) {
        guard let socketId = notification.userInfo?["socketId"] as? Int,
              let address: String = notification.userInfo?["address"] as? String,
              let port = notification.userInfo?["port"] as? Int,
              let socket = sockets[socketId] else {
            return
        }
        let data = notification.userInfo?["data"] as? Data ?? Data.init()

        try? socket.send(data, address: address, port: port)
    }
}

File: android/src/main/java/com/svend/plugins/udp/socket/UdpSocketPlugin.java
package com.svend.plugins.udp.socket;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Base64;
import android.util.Log;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.LinkedBlockingQueue;

@CapacitorPlugin(name = "UdpSocket")
public class UdpSocketPlugin extends Plugin {

    private Map<Integer, UdpSocket> sockets = new ConcurrentHashMap<>();
    private BlockingQueue<SelectorMessage> selectorMessages = new LinkedBlockingQueue<>();
    private int nextSocket = 0;
    private Selector selector;
    private SelectorThread selectorThread;

    private BroadcastReceiver dataForwardReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            int socketId = intent.getIntExtra("socketId", -1);
            String address = intent.getStringExtra("address");
            int port = intent.getIntExtra("port", -1);
            byte[] data = intent.getByteArrayExtra("data");
            try {
                UdpSocket socket = obtainSocket(socketId);
                if (!socket.isBound) throw new Exception("Not bound yet");
                socket.addSendPacket(address, port, data, null);
                addSelectorMessage(socket, SelectorMessageType.SO_ADD_WRITE_INTEREST, null);
            } catch (Exception ignored) {}
        }
    };

    @Override
    protected void handleOnStart() {
        startSelectorThread();
        LocalBroadcastManager.getInstance(getContext()).registerReceiver(dataForwardReceiver, new IntentFilter("capacitor-udp-forward"));
    }

    @Override
    protected void handleOnStop() {
        Log.i("lifecycle", "stop");
        stopSelectorThread();
    }

    @Override
    protected void handleOnRestart() {
        Log.i("lifecycle", "restart");
        startSelectorThread();
    }

    @PluginMethod
    public void create(PluginCall call) {
        try {
            JSObject properties = call.getObject("properties");
            UdpSocket socket = new UdpSocket(nextSocket++, properties);
            sockets.put(socket.getSocketId(), socket);
            socket.setReceiveCallback(
                new ReceiveCallback() {
                    @Override
                    public void receive(byte[] data, int socketId, String address, int port) {
                        sendReceiveEvent(data, socketId, address, port);
                    }

                    @Override
                    public void receiveError(int code, String message) {
                        sendReceiveErrorEvent(code, message);
                    }
                }
            );

            JSObject ret = new JSObject();
            ret.put("socketId", socket.getSocketId());
            ret.put("ipv4", socket.ipv4Address.getHostAddress());
            String ipv6 = socket.ipv6Address.getHostAddress();
            if (ipv6 != null) {
                int ip6InterfaceIndex = ipv6.indexOf("%");
                if (ip6InterfaceIndex > 0) {
                    ret.put("ipv6", ipv6.substring(0, ip6InterfaceIndex));
                } else {
                    ret.put("ipv6", ipv6);
                }
            }

            call.resolve(ret);
        } catch (Exception e) {
            call.reject("create error", e);
        }
    }

    @PluginMethod
    public void update(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            JSObject properties = call.getObject("properties");
            UdpSocket socket = obtainSocket(socketId);
            socket.setProperties(properties);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setPaused(PluginCall call) {
        int socketId = call.getInt("socketId");
        boolean paused = call.getBoolean("paused");
        try {
            UdpSocket socket = obtainSocket(socketId);
            socket.setPaused(paused);
            if (paused) {
                // Read interest will be removed when socket is readable on selector thread.
                call.resolve();
            } else {
                addSelectorMessage(socket, SelectorMessageType.SO_ADD_READ_INTEREST, call);
            }
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void bind(PluginCall call) {
        int socketId = call.getInt("socketId");
        String address = call.getString("address");
        int port = call.getInt("port");
        try {
            UdpSocket socket = obtainSocket(socketId);
            socket.bind(address, port);
            addSelectorMessage(socket, SelectorMessageType.SO_BIND, call);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void send(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            String address = call.getString("address");
            int port = call.getInt("port");
            String bufferString = call.getString("buffer");
            byte[] data = Base64.decode(bufferString, Base64.DEFAULT);
            UdpSocket socket = obtainSocket(socketId);
            if (!socket.isBound) throw new Exception("Not bound yet");
            socket.addSendPacket(address, port, data, call);
            addSelectorMessage(socket, SelectorMessageType.SO_ADD_WRITE_INTEREST, null);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void closeAllSockets(PluginCall call) {
        try {
            for (UdpSocket socket : sockets.values()) {
                addSelectorMessage(socket, SelectorMessageType.SO_CLOSE, null);
            }
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void close(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            UdpSocket socket = obtainSocket(socketId);
            addSelectorMessage(socket, SelectorMessageType.SO_CLOSE, call);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void getInfo(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            UdpSocket socket = obtainSocket(socketId);
            call.resolve(socket.getInfo());
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void getSockets(PluginCall call) {
        try {
            JSArray results = new JSArray();
            for (UdpSocket socket : sockets.values()) {
                results.put(socket.getInfo());
            }
            JSObject ret = new JSObject();
            ret.put("sockets", results);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void joinGroup(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            String address = call.getString("address");
            UdpSocket socket = obtainSocket(socketId);
            socket.joinGroup(address);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void leaveGroup(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            String address = call.getString("address");
            UdpSocket socket = obtainSocket(socketId);
            socket.leaveGroup(address);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setMulticastTimeToLive(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            int ttl = call.getInt("ttl");
            UdpSocket socket = obtainSocket(socketId);
            socket.setMulticastTimeToLive(ttl);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setBroadcast(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            boolean enabled = call.getBoolean("enabled");
            UdpSocket socket = obtainSocket(socketId);
            socket.setBroadcast(enabled);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setMulticastLoopbackMode(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            boolean enabled = call.getBoolean("enabled");
            UdpSocket socket = obtainSocket(socketId);
            socket.setMulticastLoopbackMode(enabled, call);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void getJoinedGroups(PluginCall call) {
        try {
            int socketId = call.getInt("socketId");
            UdpSocket socket = obtainSocket(socketId);

            JSArray results = new JSArray(socket.getJoinedGroups());
            JSObject ret = new JSObject();
            ret.put("groups", results);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    private void sendReceiveErrorEvent(int code, String message) {
        JSObject error = new JSObject();
        try {
            error.put("message", message);
            error.put("resultCode", code);
            notifyListeners("receiveError", error, false);
        } catch (Exception ignored) {}
    }

    private UdpSocket obtainSocket(int socketId) throws Exception {
        UdpSocket socket = sockets.get(socketId);
        if (socket == null) {
            throw new Exception("No socket with socketId " + socketId);
        }
        return socket;
    }

    // This is a synchronized method because regular read and multicast read on
    // different threads, and we need to send data and metadata in serial in order
    // to decode the receive event correctly. Alternatively, we can send Multipart
    // messages.
    private synchronized void sendReceiveEvent(byte[] data, int socketId, String address, int port) {
        JSObject ret = new JSObject();
        try {
            ret.put("socketId", socketId);
            int ip6InterfaceIndex = address.indexOf("%");
            if (ip6InterfaceIndex > 0) {
                ret.put("remoteAddress", address.substring(0, ip6InterfaceIndex));
            } else {
                ret.put("remoteAddress", address);
            }
            ret.put("remotePort", port);
            String bufferString = new String(Base64.encode(data, Base64.DEFAULT));
            ret.put("buffer", bufferString);
            notifyListeners("receive", ret, false);
        } catch (Exception ignored) {}
    }

    private void startSelectorThread() {
        if (selectorThread != null) return;
        selectorThread = new SelectorThread(selectorMessages, sockets);
        selectorThread.start();
    }

    private void stopSelectorThread() {
        if (selectorThread == null) return;
        addSelectorMessage(null, SelectorMessageType.T_STOP, null);
        try {
            selectorThread.join();
            selectorThread = null;
        } catch (InterruptedException ignored) {}
    }

    private void addSelectorMessage(UdpSocket socket, SelectorMessageType type, PluginCall call) {
        try {
            selectorMessages.put(new SelectorMessage(socket, type, call));
            if (selector != null) selector.wakeup();
        } catch (InterruptedException ignored) {}
    }

    private enum SelectorMessageType {
        SO_BIND,
        SO_CLOSE,
        SO_ADD_READ_INTEREST,
        SO_ADD_WRITE_INTEREST,
        T_STOP
    }

    private class SelectorMessage {

        final UdpSocket socket;
        final SelectorMessageType type;
        final PluginCall call;

        SelectorMessage(UdpSocket socket, SelectorMessageType type, PluginCall call) {
            this.socket = socket;
            this.type = type;
            this.call = call;
        }
    }

    private class SelectorThread extends Thread {

        private BlockingQueue<SelectorMessage> selectorMessages;
        private Map<Integer, UdpSocket> sockets;
        private boolean running = true;

        SelectorThread(BlockingQueue<SelectorMessage> selectorMessages, Map<Integer, UdpSocket> sockets) {
            this.selectorMessages = selectorMessages;
            this.sockets = sockets;
        }

        private void processPendingMessages() {
            while (selectorMessages.peek() != null) {
                SelectorMessage msg = null;
                try {
                    msg = selectorMessages.take();
                    switch (msg.type) {
                        case SO_BIND -> {
                            msg.socket.register(selector, SelectionKey.OP_READ);
                            msg.socket.isBound = true;
                        }
                        case SO_CLOSE -> {
                            msg.socket.close();
                            sockets.remove(msg.socket.getSocketId());
                        }
                        case SO_ADD_READ_INTEREST -> msg.socket.addInterestSet(SelectionKey.OP_READ);
                        case SO_ADD_WRITE_INTEREST -> msg.socket.addInterestSet(SelectionKey.OP_WRITE);
                        case T_STOP -> running = false;
                    }

                    if (msg.call != null) msg.call.resolve();
                } catch (InterruptedException ignored) {} catch (IOException e) {
                    if (msg.call != null) {
                        msg.call.reject(e.getMessage());
                    }
                }
            }
        }

        public void run() {
            try {
                if (selector == null) selector = Selector.open();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }

            // process possible messages that send during openning the selector
            // before select.
            processPendingMessages();

            Iterator<SelectionKey> it;

            while (running) {
                try {
                    selector.select();
                } catch (IOException e) {
                    continue;
                }

                it = selector.selectedKeys().iterator();

                while (it.hasNext()) {
                    SelectionKey key = it.next();
                    it.remove();

                    if (!key.isValid()) {
                        continue;
                    }

                    UdpSocket socket = (UdpSocket) key.attachment();

                    if (key.isReadable()) {
                        socket.read();
                    }

                    if (key.isWritable()) {
                        socket.dequeueSend();
                    }
                } // while next

                processPendingMessages();
            }
        }
    }
}
